### PR TITLE
mpi/test: remove dtpools static library dependency

### PR DESCRIPTION
**Pull request by gcongiu**
## Pull Request Description

Currently `make testing` fails if the library is configured with `--disable-static` as the test suite relies on `libdtpools.a` to compile the tests, see https://github.com/pmodels/mpich/issues/3844. The reason for using `libdtpools.a` explicitly instead of `-ldtpools` is that the test suite does not employee `libtool`, instead it compiles all the objects files and then places them into every single test. This PR fixes the issue by adding `libtool` support in the test suite, turning `dtpools` into a convenience library and linking it to tests using `libtool`.

## Expected Performance Changes
None

## Known Issues
None

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [x] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

File: test/mpi/Makefile.mtest [hunk 1]
```diff
@@ -19,16 +19,21 @@ AM_CPPFLAGS = -I$(top_builddir)/include -I$(top_srcdir)/include
 LDADD = $(top_builddir)/util/mtest.$(OBJEXT) $(top_builddir)/util/mtest_common.$(OBJEXT)
 
 # Add libdtpools support
+DTPOBJS = $(top_builddir)/dtpools/src/.libs/libdtpools_la-dtpools.o             \
+          $(top_builddir)/dtpools/src/.libs/libdtpools_la-dtpools_attr.o        \
+          $(top_builddir)/dtpools/src/.libs/libdtpools_la-dtpools_desc.o        \
+          $(top_builddir)/dtpools/src/.libs/libdtpools_la-dtpools_init_verify.o \
+          $(top_builddir)/dtpools/src/.libs/libdtpools_la-dtpools_misc.o
```
Comment (pavanbalaji) [inline test/mpi/Makefile.mtest]: I don't think we should be relying on `.libs` content.  That's internal libtools stuff, and the naming can change from version to version.  I think the autotools-intended way to do this is to make dtpools a convenience library and have the tests depend on it.

Comment (gcongiu) [reply test/mpi/Makefile.mtest]: I know and I agree with you this is not an ideal solution. I remember trying to link dtpools as external library in the first dtpools implementation but it didn't work. I will try again and see if I can make that work.

Comment (hzhou) [reply test/mpi/Makefile.mtest]: `.libs` is for objects built with `PIC` that is for shared library. I believe the static objects are directly in the `src` directory. Therefore, when it is configured with static only, those `.libs` objects won't be available.

Per [libtool manual](https://www.gnu.org/software/libtool/manual/html_node/Linking-executables.html#Linking-executables), Directly linking with `libdtpools.la` will create executable in `.libs` and the executable become a shell script wrapper. It is supposed to work (as runnable), but each will incur overhead (nearly half second according to the example). I think for testing purpose, we probably should prefer testing the binary directly. Then we need install dtpools before running the tests somehow.

Comment (hzhou) [reply test/mpi/Makefile.mtest]: A compromise solution maybe to intercept the configure options to `test/mpi` and always configure `dtpools` `static-only`. That way, we can leave `dtpools` intact, i.e. still fully supports static or dynamic build if someone choose to use `dtpools` independently. We also simplify the testsuite build -- always have dtpools statically linked in. Anyway, when user configure mpich with `--disable-static`, they really only meant that option for `mpich`, not testsuite. So the solution of isolating the option of `testsuite` from that of `mpich` better reflects user's intention.

Comment (raffenet) [reply test/mpi/Makefile.mtest]: Always building dtpools statically for internal use sounds like a reasonable compromise to me.

Comment (hzhou) [reply test/mpi/Makefile.mtest]: > I remember trying to link dtpools as external library in the first dtpools implementation but it didn't work. 

Chatting with Ken and realize the problem is obvious: to compile tests without installing `dtpool` only works if we build tests with `libtool` wrapper. Since we are using `mpicc` etc to build tests, that won't work.

File: configure.ac [hunk 2]
```diff
@@ -5371,7 +5371,9 @@ PAC_GET_EXENAME("mpifort", MPIFORT_NAME)
 PAC_GET_EXENAME("mpif90", MPIF90_NAME)
 PAC_GET_EXENAME("mpif77", MPIF77_NAME)
 
+AC_CONFIG_SUBDIRS([test/mpi/dtpools])
```
Comment (gcongiu) [inline configure.ac]: @pavanbalaji a possible workaround is to build `dtpools` from main mpich and install it as external package. Then using it in the test suite is just a matter of adding the right `CFLAGS` and `LDFLAGS`. This is just an attempt but I have the suspect you won't like it :)

File: test/mpi/Makefile.mtest (hunk 1 above)
Comment (gcongiu) [reply test/mpi/Makefile.mtest]: > A compromise solution maybe to intercept the configure options to `test/mpi` and always configure `dtpools` `static-only`. That way, we can leave `dtpools` intact, i.e. still fully supports static or dynamic build if someone choose to use `dtpools` independently. We also simplify the testsuite build -- always have dtpools statically linked in. Anyway, when user configure mpich with `--disable-static`, they really only meant that option for `mpich`, not testsuite. So the solution of isolating the option of `testsuite` from that of `mpich` better reflects user's intention, as well as fits our future direction (of making test suite independent)

I just saw this comment, yes that might be an option. I think Ken was also suggesting a similar approach.

**Comment by gcongiu** [comment]
@pavanbalaji @raffenet @hzhou this is another option using `AX_SUBDIRS_CONFIGURE` which allows to explicitly filter out some configure option. This option is the less invasive as it does not require any change in the Makefile inside `test/mpi`

**Comment by hzhou** [comment]
test:mpich/custom --disable-static

**Comment by raffenet** [comment]
I'm not sure we are able to include `AX_SUBDIRS_CONFIGURE` in MPICH. It does not contain the unlimited license like other `AX_*` macros we have used in the past.

**Comment by gcongiu** [comment]
> I'm not sure we are able to include `AX_SUBDIRS_CONFIGURE` in MPICH. It does not contain the unlimited license like other `AX_*` macros we have used in the past.

https://github.com/gitGNU/gnu_autoconf-archive/tree/cfa14e9ac5d874abfc19e5d33eb7a1bbb9c4e899 says these scripts can be reused freely

**Comment by raffenet** [comment]
@gcongiu the can be reused freely under the terms of GPLv3. MPICH is not compatible with GPLv3. Some macros grant an additional unlimited license. This one does not appear to do so.

**Comment by gcongiu** [comment]
@raffenet got it

**Comment by hzhou** [comment]
What a pity, the custom tests show that it works perfectly!

**Comment by pavanbalaji** [comment]
Why is my suggestion above not possible?

File: test/mpi/Makefile.mtest (hunk 1 above)
Comment (gcongiu) [reply test/mpi/Makefile.mtest]: @pavanbalaji I think this is closer to what I had implemented in patch #2 except we don't want to expose dtpools externally

**Comment by pavanbalaji** [comment]
I just talked to @gcongiu.  The main problem seems to be that the test suite wasn't using libtool for some reason.  If we add the `LT_INIT` and whatever else is needed in `configure.ac`, all of these problems would just go away and we won't need to do any of these hacks of disable static libraries or tweaking around CFLAGS/LDFLAGS.  `libtool` will take care of all those combinations.

**Comment by gcongiu** [comment]
test:mpich/custom --disable-static

**Comment by gcongiu** [comment]
@pavanbalaji this should work. I have tested it locally now running it on jenkins.

File: test/mpi/Makefile.mtest [hunk 3]
```diff
@@ -14,25 +14,20 @@
 ## The goal is to reduce the amount of boilerplate that needs to exist in each
 ## Makefile.am
 
-# AM_CPPFLAGS are used for C++ code as well
-AM_CPPFLAGS = -I$(top_builddir)/include -I$(top_srcdir)/include
-LDADD = $(top_builddir)/util/mtest.$(OBJEXT) $(top_builddir)/util/mtest_common.$(OBJEXT)
+AM_CPPFLAGS = -I$(top_srcdir)/include -I$(top_srcdir)/dtpools/include
+LDADD = libmtest.la $(top_builddir)/dtpools/src/.libs/libdtpools.la
```
Comment (pavanbalaji) [inline test/mpi/Makefile.mtest]: This doesn't look right.  I don't think you should be assuming anything in `.libs`.  I think the libtool intended way to do this is to add these dependency libraries to `executable_DEPENDENCIES`.

Comment (pavanbalaji) [reply test/mpi/Makefile.mtest]: Sorry, I should have said `executable_LDADD` to add the libraries + `executable_DEPENDENCIES` for dependency tracking.

Comment (pavanbalaji) [reply test/mpi/Makefile.mtest]: For example, in Hydra, we do this:

```
mpiexec_hydra_LDADD = -lpm -lhydra $(external_libs)
mpiexec_hydra_DEPENDENCIES = libpm.la libhydra.la
```

**Comment by gcongiu** [comment]
@pavanbalaji I had to convert `dtpools` into a convenience library as initially it was actually thought to be installed as additional package. Testing again to make sure this works

test:mpich/custom --disable-static

**Comment by hzhou** [comment]
@gcongiu Do you have to turn `mtest` into a convenience library in this PR? If possible, I think turning `mtest` into a convenience library should be a separate PR.

**Comment by gcongiu** [comment]
@hzhou yes you are right I was actually going to separate the dtpools conversion into convenience library into a separate PR and leave mtest libtool integration into this one. Unfortunately I am still not able to trigger dtpools compilation when `make testing`. Everything works as long as `make` is invoked first.

**Comment by gcongiu** [comment]
test:mpich/custom --disable-static

**Comment by gcongiu** [comment]
test:mpich/custom --disable-static

File: test/mpi/Makefile.mtest [hunk 4]
```diff
@@ -20,15 +20,15 @@ LDADD = $(top_builddir)/util/mtest.$(OBJEXT) $(top_builddir)/util/mtest_common.$
...
+LDADD += $(top_builddir)/dtpools/src/libdtpools.la
 
 $(top_builddir)/util/mtest.$(OBJEXT): $(top_srcdir)/util/mtest.c
 	(cd $(top_builddir)/util && $(MAKE) mtest.$(OBJEXT))
 
 $(top_builddir)/util/mtest_common.$(OBJEXT): $(top_srcdir)/util/mtest_common.c
 	(cd $(top_builddir)/util && $(MAKE) mtest_common.$(OBJEXT))
 
-$(top_builddir)/dtpools/src/.libs/libdtpools.a: $(top_srcdir)/dtpools/src/*.c
+$(top_builddir)/dtpools/src/libdtpools.la:
```
Comment (gcongiu) [inline test/mpi/Makefile.mtest]: @pavanbalaji I had to leave this or else `make testing` won't be able to trigger `dtpools` compilation. Making this work more seemingly with `libtool` will require more work on the testsuite (which includes the conversion of `mtest` into a convenience library). As @hzhou has suggested, this should be a separate PR thus I am not addressing this here as I was initially trying to do.

**Comment by raffenet** [comment]
Technically this looks fine. I think the decision to use `libtool` to link all the tests is already decided. I have no objection then to this PR.

**Comment by hzhou** [comment]
test:jenkins/ch3/sock
test:jenkins/ch4/ucx

**Comment by gcongiu** [comment]
@pavanbalaji are you also happy with the PR?
